llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp

```diff
@@ -1250,6 +1265,158 @@ void AddressSanitizerPass::printPipeline(
+              F.getName().equals("_Z14_get_num_groupsj"))
+            continue;
+        }
+
+      } else {
+        if (F.isDeclaration() || F.getName().starts_with("__sycl_") ||
+            F.getName().starts_with("__spirv") ||
+            F.getFunctionType()->isVarArg())
+          continue;
+      }
+      SpirFixupFuncs.emplace_back(&F);
+    }
+  }
+
+  SmallVector<std::pair<Function *, Function *>> SpirFuncs;
+  const int LongSize = M.getDataLayout().getPointerSizeInBits();
+  auto *IntptrTy = Type::getIntNTy(M.getContext(), LongSize);
+
+  for (auto *F : SpirFixupFuncs) {
+    SmallVector<Type *, 16> Types;
+    for (Function::const_arg_iterator I = F->arg_begin(), E = F->arg_end();
+         I != E; ++I) {
+      Types.push_back(I->getType());
+    }
+
+    // New Argument
+    Types.push_back(IntptrTy);
+
+    FunctionType *NewFTy = FunctionType::get(F->getReturnType(), Types, false);
+
+    std::string OrigFuncName = F->getName().str();
+    F->setName(OrigFuncName + "_del");
+
+    Function *NewF =
+        Function::Create(NewFTy, F->getLinkage(), OrigFuncName, F->getParent());
+    NewF->copyAttributesFrom(F);
+    NewF->copyMetadata(F, 0);
+    NewF->setCallingConv(F->getCallingConv());
+    NewF->setDSOLocal(F->isDSOLocal());
+
+    // Set original arguments' names.
+    Function::arg_iterator NewI = NewF->arg_begin();
+    for (Function::const_arg_iterator I = F->arg_begin(), E = F->arg_end();
+         I != E; ++I, ++NewI) {
+      NewI->setName(I->getName());
+    }
+    // New argument name
+    NewI->setName("__asan_launch");
+
+    NewF->splice(NewF->begin(), F);
+    assert(F->isDeclaration() &&
+           "splice does not work, original function body is not empty!");
+
+    NewF->setSubprogram(F->getSubprogram());
+
+    NewF->setComdat(F->getComdat());
+    F->setComdat(nullptr);
+
+    F->deleteBody();
+
+    for (Function::arg_iterator I = F->arg_begin(), E = F->arg_end(),
+                                NI = NewF->arg_begin();
+         I != E; ++I, ++NI) {
+      I->replaceAllUsesWith(&*NI);
+    }
+
+    // Fixup metadata
+    IRBuilder<> Builder(M.getContext());
+
+    auto FixupMetadata = [&NewF](StringRef MDName, Constant *NewV) {
+      auto *Node = NewF->getMetadata(MDName);
+      if (!Node)
+        return;
+      SmallVector<Metadata *, 8> NewMD;
+      for (unsigned I = 0; I < Node->getNumOperands(); ++I) {
+        NewMD.emplace_back(Node->getOperand(I));
+      }
+      NewMD.emplace_back(ConstantAsMetadata::get(NewV));
+      NewF->setMetadata(MDName, llvm::MDNode::get(NewF->getContext(), NewMD));
+    };
+
+    FixupMetadata("kernel_arg_buffer_location", Builder.getInt32(-1));
+    FixupMetadata("kernel_arg_runtime_aligned", Builder.getFalse());
+    FixupMetadata("kernel_arg_exclusive_ptr", Builder.getFalse());
+
+    if (F->getCallingConv() == CallingConv::SPIR_KERNEL)
+      F->removeFromParent();
+    else
+      SpirFuncs.emplace_back(F, NewF);
```

Comment:
can a sycl kernel be called from another sycl kernel as well?

Reply:
I'm not 100% sure about this, I assume sycl kernel can't call another sycl kernel.
This is the answer from intelgpt ...
```
As of the SYCL 2020 specification, SYCL does not natively support invoking one kernel directly from another kernel, a concept often referred to as dynamic parallelism in the context of some other parallel computing platforms like CUDA.

In SYCL, the programming model is generally centered around the submission of kernels from the host code to the device. The kernels are enqueued onto command queues, and the SYCL runtime handles their scheduling and execution. This model does not inherently support kernel-to-kernel launches as one might see with CUDA's dynamic parallelism feature.
```

Reply:
ok

Reply:
> I'm not 100% sure about this, I assume sycl kernel can't call another sycl kernel. This is the answer from IntelGPU ...
> 
> > As of the SYCL 2020 specification, SYCL does not natively support invoking one kernel directly from another kernel, a concept often referred to as dynamic parallelism in the context of some other parallel computing platforms like CUDA.
> > In SYCL, the programming model is generally centered around the submission of kernels from the host code to the device. The kernels are enqueued onto command queues, and the SYCL runtime handles their scheduling and execution. This model does not inherently support kernel-to-kernel launches as one might see with CUDA's dynamic parallelism feature.

I don't think it applies to the @wenju-he's question as his question is not related to the dynamic parallelism. According to my understanding, the question is "can a function with spir_kernel calling convention call another function with spir_kernel calling convention"?
This is not prohibited by SYCL or SPIR-V standards. It might be tricky to create a reproducer for that case using existing DPC++ implementation, but it's not impossible.

BTW, who is IntelGPU?

Reply:
Thanks for correcting me. I'll handle this situation as well.

"IntelGPU" -> "IntelGPT", typo. 
I found it's hard to search from SYCL spec or internet, so I tried to ask ChatGPT for this.

Reply:
> This is not prohibited by SYCL or SPIR-V standards. It might be tricky to create a reproducer for that case using existing DPC++ implementation, but it's not impossible.

Hi @bader, can you help to craft one example? 
I tried to use function object but failed, and device code can't use function pointer either.

Reply:
Now we only extend kernels' argument, if we can't call kernel in device code, then I needn't to fix the users of kernels.